flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/allocator/StateLocalitySlotAssigner.java

```diff
@@ -198,4 +195,18 @@ public Collection<AllocationScore> calculateScore(
                 .map(e -> new AllocationScore(group.getId(), e.getKey(), e.getValue()))
                 .collect(Collectors.toList());
     }
+
+    private static long estimateSize(
+            KeyGroupRange newRange, VertexAllocationInformation allocation) {
+        KeyGroupRange oldRange = allocation.getKeyGroupRange();
+        if (allocation.stateSizeInBytes * oldRange.getNumberOfKeyGroups() == 0) {
+            return 0L;
+        }
+        // round up to 1
+        long keyGroupSize =
```

Comment:
So basically this assumes all local key groups have the same state size. Do you think this is a significant improvement over just looking at the number of key groups? I know it is at least a bit more precise...

Reply:
Yes, I think the state size of different subtasks might vary significantly even if the number of key groups is the same, because of the hot keys/groups.
I've also remember some patterns of working with state where a single subtask 0 handles all the state.

Besides that, with slot sharing, single slot can be reserved for multiple tasks, having completely different state size and parallelism. Without state size estimation, stateless high-parallelism tasks might outweigh state-heavy low-parallelism tasks while scoring.